### PR TITLE
blastdamage

### PR DESCRIPTION
**Pull request by LightRobin**
adds blastDamage() function, attacker is automatically set to instance.player so that the player calling the function always shows as the killer, (easy to track down abuse)

### CODE REVIEW

**Comment by adamnejm** [comment]
I feel like this needs permission check, maybe even a delay convar.
Also wouldn't it be possible to trick the kill feed by letting user specify the `inflictor`, or does that rely solely on the `attacker`?

**Comment by LightRobin** [comment]
i believe it is solely based on the attacker, i cooldown might be good but the reason i though of adding this was to make prop engines that use the blast as propulsion for the pistons, adding a delay would break that possibility

**Comment by LightRobin** [comment]
what would the permission check check for?, you cant specify which entities the blast will affect

**Comment by adamnejm** [comment]
> what would the permission check check for?

For server owners to enable / disable / limit this feature.

File: lua/starfall/libs_sh/builtins.lua [hunk 1]
```diff
@@ -528,6 +528,19 @@ if SERVER then
...
+	--- Applies explosion damage to all entities in the specified radius
+	-- @server
+	-- @param inflictor The entity that caused the damage
+	-- @param damageOrigin The center of the explosion
+	-- @param damageRadius The radius in which entities will be damaged
+	-- @param damage The amount of damage to be applied
+	function builtins_library.blastDamage(inflictor, damageOrigin, damageRadius, damage)
+		inflictor, damageOrigin = ounwrap(inflictor), vunwrap(damageOrigin)
+		checkluatype(inflictor, TYPE_ENTITY)
+		checkluatype(damageOrigin, TYPE_VECTOR)
```
Comment (thegrb93) [inline lua/starfall/libs_sh/builtins.lua:541]: The checkluatype isn't needed since unwrap will do that for you. Use eunwrap instead of ounwrap though.

**Review by thegrb93 — CHANGES_REQUESTED** [review]

**Comment by thegrb93** [comment]
The last thing I'm worried about is abuse and many servers have wire explosives turned off specifically due to players teleporting explosives around (dumb I know).

**Comment by LightRobin** [comment]
ye, i think it needs a permission that the owner can set (everyone, admin only, noone) i tried adding it earlier but simply ended up with lua errors, heres the code i used 
`registerPrivilege("blast.create", "Blast damage", "Allows the user to create explosions", { server = { default = 1 } })`
`(in the function) checkpermission(instance, nil, "blast.create")`

was getting this error :
`addons/starfall/lua/starfall/permissions/core.lua:69: attempt to call a nil value`
`stack traceback:`
`    addons/starfall/lua/starfall/permissions/core.lua:69: in function 'checkpermission'`

**Comment by Vurv78** [comment]
You weren't trying to create the permission inside the instance, were you?

**Comment by thegrb93** [comment]
> 
> 
> ye, i think it needs a permission that the owner can set (everyone, admin only, noone) i tried adding it earlier but simply ended up with lua errors, heres the code i used
> `registerPrivilege("blast.create", "Blast damage", "Allows the user to create explosions", { server = { default = 1 } })`
> `(in the function) checkpermission(instance, nil, "blast.create")`
> 
> was getting this error :
> `addons/starfall/lua/starfall/permissions/core.lua:69: attempt to call a nil value`
> `stack traceback:`
> ` addons/starfall/lua/starfall/permissions/core.lua:69: in function 'checkpermission'`

usergroups = { default = 1 }

**Comment by LightRobin** [comment]
> You weren't trying to create the permission inside the instance, were you?

Nope, just under the other permissions

**Comment by LightRobin** [comment]
> usergroups = { default = 1 }
still getting the same error

**Comment by thegrb93** [comment]
> 
> 
> > usergroups = { default = 1 }
> > still getting the same error

Looks like the problem is you just need to reload the map since its a new permission and hasn't been set up.

**Comment by thegrb93** [comment]
Moved to game lib and rebased to resolve conflicts. Let me know if changes look fine.

**Comment by LightRobin** [comment]
looks good to me

**Pull request merged**